### PR TITLE
[36384] Add job state time limit actions to batch queue

### LINKED ISSUE #36384: [Enhancement]: `aws_batch_job_queue` jobStateTimeLimitActions parameter
### Description

AWS Batch now supports a Batch Job Queue Blocked CloudWatch Event for jobs stuck in RUNNABLE state. Customers can automate actions using EventBridge to be able to take an action on these stuck jobs. Additionally, customers can configure jobStateTimeLimitActions parameter from the CreateJobQueue and UpdateJobQueue APIs to terminate the stuck job unblocking the jobs behind it within the queue. 

### Affected Resource(s) and/or Data Source(s)

* aws_batch_job_queue

### Potential Terraform Configuration

```terraform
job_state_time_limit_actions {
   reason         = "string" 
   state          = "string"            
   maxTimeSeconds = number     
   action         = "string"            
}
```

### References

- blog post: https://aws.amazon.com/about-aws/whats-new/2024/03/aws-batch-alerts-detect-jobs-runnable-state/
- sdk update: https://github.com/aws/aws-sdk-go/commit/b965d6132d94bb7314bd114e3691a3f7005c191d

### Would you like to implement a fix?

Yes

### PR DESCRIPTION
**Pull request by jchorl**
### Description
A new option has been added to batch job queues: https://aws.amazon.com/about-aws/whats-new/2024/03/aws-batch-alerts-detect-jobs-runnable-state/

This PR just adds support for the new API fields to the batch queue terraform resource.

### Relations

Closes #36384

### References

API reference: https://docs.aws.amazon.com/batch/latest/APIReference/API_JobStateTimeLimitAction.html

### Output from Acceptance Testing

I didn't run acceptance tests to avoid dangling resources in my account.

I did however use this to update job queues and it worked.

### CODE REVIEW

File: internal/service/batch/job_queue.go [hunk 1]
```diff
@@ -133,6 +133,25 @@ func (r *resourceJobQueue) Schema(ctx context.Context, request resource.SchemaRe
 				},
 			},
 		},
+		"job_state_time_limit_action": schema.ListNestedBlock{
+			CustomType: fwtypes.NewListNestedObjectTypeOf[jobStateTimeLimitAction](ctx),
+			NestedObject: schema.NestedBlockObject{
+				Attributes: map[string]schema.Attribute{
+					"action": schema.StringAttribute{
+						Required: true,
+					},
```
Comment (drewmullen) [inline internal/service/batch/job_queue.go:142]:
```suggestion
						ValidateFunc: validation.StringInSlice(ec2.JobStateTimeLimitActionsAction_Values(), false),
					},
```

File: internal/service/batch/job_queue.go [hunk 2]
```diff
@@ -133,6 +133,25 @@ func (r *resourceJobQueue) Schema(ctx context.Context, request resource.SchemaRe
...
+					"max_time_seconds": schema.Int64Attribute{
+						Required: true,
+					},
```
Comment (drewmullen) [inline internal/service/batch/job_queue.go:145]:
```suggestion
						ValidateFunc: validation.IntBetween(600, 86400),
					},
```

File: internal/service/batch/job_queue.go [hunk 3]
```diff
@@ -133,6 +133,25 @@ func (r *resourceJobQueue) Schema(ctx context.Context, request resource.SchemaRe
...
+					"reason": schema.StringAttribute{
+						Required: true,
+					},
+					"state": schema.StringAttribute{
+						Required: true,
+					},
```
Comment (drewmullen) [inline internal/service/batch/job_queue.go:151]:
```suggestion
						ValidateFunc: validation.StringInSlice(ec2.JobStateTimeLimitActionsState_Values(), false),
					},
```

File: internal/service/batch/job_queue_test.go [hunk 4]
```diff
@@ -453,6 +453,38 @@ func TestAccBatchJobQueue_tags(t *testing.T) {
...
+					resource.TestCheckResourceAttrPair(resourceName, "job_state_time_limit_action.0", "job_state_time_limit_action.reason", "foobar"),
+					resource.TestCheckResourceAttrPair(resourceName, "job_state_time_limit_action.0", "job_state_time_limit_action.state", "RUNNABLE"),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
```
Comment (drewmullen) [inline internal/service/batch/job_queue_test.go:484]: - [ ] Please add a test with 2 `job_state_time_limit_action` defined 
- [ ] Please add a test where you update the `reason` and `max_time_seconds`

File: website/docs/r/batch_job_queue.html.markdown [hunk 5]
```diff
@@ -86,6 +87,13 @@ This resource supports the following arguments:
 * `compute_environment` - (Required) The Amazon Resource Name (ARN) of the compute environment.
 * `order` - (Required) The order of the compute environment. Compute environments are tried in ascending order. For example, if two compute environments are associated with a job queue, the compute environment with a lower order integer value is tried for job placement first.
 
+### job_state_time_limit_action
+
+* `action` - (Required) The action to take when a job is at the head of the job queue in the specified state for the specified period of time.
+* `max_time_seconds` - (Required) The approximate amount of time, in seconds, that must pass with the job in the specified state before the action is taken.
```
Comment (drewmullen) [inline website/docs/r/batch_job_queue.html.markdown:93]:
```suggestion
    * `job_state_time_limit_action.#.max_time_seconds` - The approximate amount of time, in seconds, that must pass with the job in the specified state before the action is taken. Valid values include integers between `600` & `86400`
```

File: website/docs/r/batch_job_queue.html.markdown [hunk 6]
```diff
@@ -86,6 +87,13 @@ This resource supports the following arguments:
...
+* `reason` - (Required) The reason to log for the action being taken.
+* `state` - (Required) The state of the job needed to trigger the action.
```
Comment (drewmullen) [inline website/docs/r/batch_job_queue.html.markdown:95]:
```suggestion
* `state` - (Required) The state of the job needed to trigger the action. Valid values include `"RUNNABLE"`.
```

File: website/docs/r/batch_job_queue.html.markdown [hunk 7]
```diff
@@ -86,6 +87,13 @@ This resource supports the following arguments:
 * `compute_environment` - (Required) The Amazon Resource Name (ARN) of the compute environment.
 * `order` - (Required) The order of the compute environment. Compute environments are tried in ascending order. For example, if two compute environments are associated with a job queue, the compute environment with a lower order integer value is tried for job placement first.
 
+### job_state_time_limit_action
+
+* `action` - (Required) The action to take when a job is at the head of the job queue in the specified state for the specified period of time.
```
Comment (drewmullen) [inline website/docs/r/batch_job_queue.html.markdown:92]:
```suggestion
* `action` - (Required) The action to take when a job is at the head of the job queue in the specified state for the specified period of time. Valid values include `"CANCEL"`
```

**Review by drewmullen — CHANGES_REQUESTED** [review]
Thanks for opening this PR. Was planning to get to that this week... appreciate you stepping up. Here's some requested changes:

Also please include a [changelog entry](https://hashicorp.github.io/terraform-provider-aws/changelog-process/)

File: internal/service/batch/job_queue.go [hunk 8]
```diff
@@ -259,6 +292,11 @@ func (r *resourceJobQueue) Update(ctx context.Context, request resource.UpdateRe
 		}
 	}
 
+	if !plan.JobStateTimeLimitAction.IsNull() && !plan.JobStateTimeLimitAction.Equal(state.JobStateTimeLimitAction) {
+		flex.Expand(ctx, plan.JobStateTimeLimitAction, &input.JobStateTimeLimitActions)
+		update = true
+	}
```
Comment (drewmullen) [inline internal/service/batch/job_queue.go:298]: can `reason` be updated?

**Comment by drewmullen** [comment]
@jchorl looks like I gave you the validation functions using the old framework and this resource is on the new framework… I can send some examples tomorrow if you have trouble

**Comment by jchorl** [comment]
> @jchorl looks like I gave you the validation functions using the old framework and this resource is on the new framework… I can send some examples tomorrow if you have trouble

You're too fast, just figuring it out locally. Will push updates soon.

**Comment by drewmullen** [comment]
I saw the notification and immediately felt bad for giving bad code lol

@jchorl

File: internal/service/batch/job_queue_test.go (hunk 4 above)
Comment (jchorl) [reply internal/service/batch/job_queue_test.go:549]: Done!

**Comment by jchorl** [comment]
I think I addressed all feedback. Unfortunately I can't test updating the `reason` this sec because there's some unrelated concurrent changes in-flux that can't be targetted around. I can try to test tomorrow.

File: internal/service/batch/job_queue_test.go [hunk 9]
```diff
@@ -453,6 +453,103 @@ func TestAccBatchJobQueue_tags(t *testing.T) {
...
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+			{
+				Config: acctest.ConfigCompose(
+					testAccJobQueueConfigBase(rName),
```
Comment (drewmullen) [inline internal/service/batch/job_queue_test.go:513]:
```suggestion
					testAccJobQueueConfig_Base(rName),
```

**Comment by ChannyClaus** [comment]
@jchorl are you still working on this PR? if you're busy with other things, happy to take over (this feature would be incredibly helpful for us 😅)

**Comment by jchorl** [comment]
> @jchorl are you still working on this PR? if you're busy with other things, happy to take over (this feature would be incredibly helpful for us 😅)

Please feel free to take it over - I haven't had a chance to figure out terraform tests, especially with all the resources they configure in your AWS acct.

I do think there is a lingering issue in this PR with state getting out of sync - i.e. if you `apply` the same config twice I think it still shows changes.

**Comment by ChannyClaus** [comment]
> > @jchorl are you still working on this PR? if you're busy with other things, happy to take over (this feature would be incredibly helpful for us 😅)
> 
> Please feel free to take it over - I haven't had a chance to figure out terraform tests, especially with all the resources they configure in your AWS acct.
> 
> I do think there is a lingering issue in this PR with state getting out of sync - i.e. if you `apply` the same config twice I think it still shows changes.

made a PR https://github.com/hashicorp/terraform-provider-aws/pull/38784 to pick this up. do you happen to remember the configuration with which you ran into the out-of-sync state issue? seems like when i make changes to the `job_state_time_limit_action` attribute, the subsequent runs of `terraform apply` shows no change.

on a related-ish note, hopefully someone will be able to take a look at the PR soon, since given the acquisition going on at my company, it's unclear how much longer i'll have access to AWS account here :.)

**Comment by jchorl** [comment]
> > > @jchorl are you still working on this PR? if you're busy with other things, happy to take over (this feature would be incredibly helpful for us 😅)
> > 
> > 
> > Please feel free to take it over - I haven't had a chance to figure out terraform tests, especially with all the resources they configure in your AWS acct.
> > I do think there is a lingering issue in this PR with state getting out of sync - i.e. if you `apply` the same config twice I think it still shows changes.
> 
> made a PR #38784 to pick this up. do you happen to remember the configuration with which you ran into the out-of-sync state issue? seems like when i make changes to the `job_state_time_limit_action` attribute, the subsequent runs of `terraform apply` shows no change.
> 
> on a related-ish note, hopefully someone will be able to take a look at the PR soon, since given the acquisition going on at my company, it's unclear how much longer i'll have access to AWS account here :.)

I just built my branch and tf-applied off that provider. Got a bunch of:
```
│ Error: Provider produced inconsistent final plan                                                                                                                                             
│                                                                                                                                                                                              
│ When expanding the plan for module.whatever.compute_env["foo"] to include new values learned so far during apply, provider               
│ "registry.terraform.io/hashicorp/aws" changed the planned action from Update to DeleteThenCreate.                                                                                            
│                                                                                                                                                                                              
│ This is a bug in the provider, which should be reported in the provider's own issue tracker. 
```

I'm not sure what this error means/why it would occur. This was off just a normal `terraform apply`. If you can toggle on/off the `job_state_time_limit_action`, tf apply, and not hit this, maybe it's fixed.

Genuine thanks for picking this up.

**Comment by ChannyClaus** [comment]
yup, tried it with 
```
terraform {
  required_providers {
	aws = {
      source  = "terraform.local/local/aws"
    }
  }
}

# Configure the AWS Provider
provider "aws" {
  region = "us-west-2"
}


resource "aws_batch_compute_environment" "sample" {
  compute_environment_name = "sample"

  compute_resources {
    max_vcpus = 16
    subnets = [
      "<redacted>",
    ] 
    security_group_ids = [
      "<redacted>",
    ]
    type = "FARGATE"
  }

  type         = "MANAGED"
}

resource "aws_batch_job_queue" "test_queue" {
  name     = "tf-test-batch-job-queue4"
  state    = "ENABLED"
  priority = 1
  compute_environment_order {
    order               = 1
    compute_environment = aws_batch_compute_environment.sample.arn
  }
  job_state_time_limit_action {
    action           = "CANCEL"
    max_time_seconds = 606
    reason           = "CAPACITY:INSUFFICIENT_INSTANCE_CAPACITY"
    state            = "RUNNABLE"
  }
}
```
i don't seem to get that error message.

and no problem! we ended up setting up a cronjob to pick up on jobs that get stuck on RUNNABLE and i'm just glad AWS got around to support this natively on their end.

**Pull request closed without merging**